### DISCUSSION TITLE
Vercel .env Variables are blank in Production and Preview

### DISCUSSION THREAD

**Discussion by surjithctly** [vercel/vercel#5015] [Help] [answered]
I was developing a simple vercel app using Github API. I have used vercel cli for local development. All went well. But when I tested in Vercel server, the app broke.

After further investigation I found that .env variable is returning blank. I tried to console.log the variable retrieved from process.env.GITHUB_TOKEN

But it returned blank in function log. 

How to fix this? Is this a Vercel Error?

My code can be found here: https://github.com/surjithctly/github-repo-access

Please help. I’m stuck.

**Comment by styfle** [comment, marked as answer]
Vercel doesn't read `.env` files for deployments, that is only for `vercel dev` convenience. 

Instead, you should use `vercel env add` command or the Dashboard on https://vercel.com to add environment variables to the cloud.

Try `vercel env --help` to see examples or view the docs here: https://vercel.com/docs/environment-variables

**Comment by surjithctly** [comment]
@styfle Yes. I know that. I have already added env variable in the settings. See:

![image](https://user-images.githubusercontent.com/1884712/89757671-8ae6fc80-db03-11ea-81c0-2a7baeab0fcf.png)

But I'm still using the `process.env `code for getting the env variable. But its returning blank. What do you think the problem might be?

```
const github_token = process.env.GITHUB_TOKEN;

console.log("token: ", github_token);
```

Result:

```
token:   
```


~Surjith

**Comment by surjithctly** [comment]
Update: 

I did a fresh testing today (put one day gap because of Sunday)

Now it seems to be working as expected. Cool. 

So my guess it it should be a Vercel .env Cache issue? What do you think?

~Surjith

**Reply by dbertella** [reply]
I have the same issue today, it was working fine up to this morning, last time I tried and now vercel build is failing because it seems like it doesn't see an environment variable. I saw is an issue that some people "fixed" it recreating the project, I'd want to avoid that if possible. 

EDIT my issue was about having different variables for prod / preview, in the old ui you weren't adding variables all together and i didn't add the one for preview. My build was failing on pull requests

**Reply by SalahAdDin** [reply]
Well, I'm getting the same problem with the preview token, it seems it does not exist even when I set up it before:
![Screen Capture_select-area_20201204004844](https://user-images.githubusercontent.com/5034215/101093616-1d883b80-35cc-11eb-882e-6b6a2150e097.jpg)

**Reply by surjithctly** [reply]
@SalahAdDin Seems Vercel issue for sure. May be try creating new issue mentioning this discussion?

**Reply by SalahAdDin** [reply]
should we?

**Reply by marcosvmds** [reply]
Same issue here guys.. 
Some news?

**Reply by Onokaev** [reply]
Same issue here. Has anyone solved this?

**Reply by styfle** [reply]
Make sure you're using the lasted version of Vercel CLI (22.0.1 at this time)

**Reply by Harold-debug** [reply]
I still got the same issue today... any solution yet guys??

**Reply by Onokaev** [reply]
Hey @Harold-debug, add the variables through the cli.
Adding the variables through the vercel dashboard sort of doesn't work sometimes.

**Reply by Harold-debug** [reply]
Thanks @Onokaev, gotta try that now

**Reply by thomas-corrado** [reply]
did this work for you? i tried doing this, but it still did not work

**Reply by amyegan** [reply]
Hi @thomas-corrado! Would you mind creating a [new discussion](https://github.com/vercel/community/discussions/new) with  details about your project? If you can share your repo or a [minimal reproducible example](https://vercel.com/support/articles/creating-a-minimal-reproducible-example), that would be very helpful. Information about your general Project Settings would also be useful to share.

**Reply by thomas-corrado** [reply]
Hi @amyegan ! I was actually able to resolve the issue, thank you so much. I thank you for taking your time and consideration to address my issue.

**Comment by Onokaev** [comment]
Thanks. Let me try that

**Comment by Onokaev** [comment]
Thanks. It worked

**Comment by oscarvz** [comment]
It seems the issue is still present. I tried multiple approaches (dashboard, CLI) but the variables turn up `undefined` in production.
I even opened new projects to verify the problem, and the problem occurs on every project that I have running.
Is there anything I can do, or should I open an issue and/or contact support?

**Comment by RodRitter** [comment]
I'm getting the same issue. I started a new application today and added my env variables to the dashboard. The app isn't detecting them.

Caveat: not sure if it's related but I added my own domain.

**Reply by aausek** [reply]
I'm seeing the same issue in my NextJS build. Even prefixed the variable in the UI with NEXT_PUBLIC_ as recommended on NextJs docs but Vercel is still not reading it.

**Comment by Onokaev** [comment]
Use the CLI to add environment variables

**Reply by aausek** [reply]
Is there a difference between using the CLI or the UI? Below is how I'm using it on localhost which works just fine. I've now tried Vercel and Github Actions and can't get either to work. 
`myHeaders.append(
    "Authorization",
    process.env.NEXT_PUBLIC_TOKEN
  );`

**Reply by SalahAdDin** [reply]
> Is there a difference between using the CLI or the UI? Below is how I'm using it on localhost which works just fine. I've now tried Vercel and Github Actions and can't get either to work. `myHeaders.append( "Authorization", process.env.NEXT_PUBLIC_TOKEN );`

Good question!

**Comment by ETARAZ** [comment]
I struggle with this issue before and I have created the following script with NodeJS using `vercel --build-env KEY1=value1 --build-env KEY2=value2` you can visit [Vercel docs](https://vercel.com/docs/cli#commands/overview/unique-options/build-env) to learn more.

```
const { exec } = require('node:child_process')
const fs = require('fs');
const readline = require('readline');


function deployWithEnv(cmd){
  exec(cmd, (err, output) => {
    // once the command has completed, the callback function is called
    if (err) {
        // log and return if we encounter an error
        console.error("could not execute command: ", err)
        return
    }
    // log the output received from the command
    console.log(output)
})
}


async function processLineByLine() {
  const fileStream = fs.createReadStream(".env");
  const rl = readline.createInterface({
    input: fileStream,
    crlfDelay: Infinity
  });
  const keyValues=[]
  // Note: we use the crlfDelay option to recognize all instances of CR LF
  // ('\r\n') in .env as a single line break.

  for await (const line of rl) {
    const li=line.trim()
    keyValues.push(`--build-env ${li}`)
  }

  return keyValues;
}
async function run(){
  const result=await processLineByLine()
  deployWithEnv(`vercel ${result.join(" ")}`)
}
run()
```

**Reply by pushkarsingh019** [reply]
adding environment variables should not be be this complex

**Reply by pushkarsingh019** [reply]
appreciate the help though, anyone has heard from the official vercel team?

**Comment by ghost** [comment]
Ensure that after adding a new Environment Variable, you have redeployed your application. Environment Variables are made available to the app on each deployment. If you attempt to read an Environment Variable that was added after the deployment was made, it's value will be undefined.

**Comment by austinschrader** [comment]
This bug still exists. I've redeployed my application, etc.

This bug is kind of impressive because it has to be the most basic thing that Vercel is supposed to provide :x

**Reply by amyegan** [reply]
Sorry you're having trouble with environment variables @austinschrader. 😞 

Would you mind creating a [new discussion](https://github.com/vercel/community/discussions/new) with details about your project? If you can share your repo or a [minimal reproducible example](https://vercel.com/support/articles/creating-a-minimal-reproducible-example), that is usually helpful for identifying the source of the problem. Information about your general Project Settings useful for debugging.

I'm curious to figure out if this is a strange interaction between Vercel and a particular framework or bundler that we need to address. Maybe there is a quick solution available. 🙏

**Reply by pushkarsingh019** [reply]
I am using create-react-app and facing the same issue, @amyegan any solutions you would recommend for this?

**Reply by amyegan** [reply]
Someone else shared the answer in https://github.com/vercel/community/discussions/1338#discussioncomment-4692590. Linking for better visibility 🙂

**Comment by robertn702** [comment]
I'm also running into the same issue on my NextJS projects in Vercel. The issue seems to be intermittent and and usually be fixed by redeploying, but that's not ideal. An example of my usage is checking the `NEXT_PUBLIC_VERCEL_ENV` variable to conditionally inject an analytics script in production. Sometimes the variable is `undefined` so the analytics script is not injected. Redeploying usually fixes the issue, but by then we have gaps in our analytics.

**Reply by FeMaffezzolli** [reply]
Have you created the env using the CLI?
